### ISSUE TITLE
Add version to _config.yml and warn if a different version is used on build

### ISSUE THREAD

**Issue by GeorgePhillips** [jekyll/jekyll#4539]
If we could add a version field to _config.yml this would save debugging time when using multiple versions of Jekyll. An example _config.yml would be:

```
name: Version Test
author: george@cloudcannon.com
version: 3.0.3
```

If you ran `jekyll serve` and it ran anything other than 3.0.3 it would output:

```bash
$ jekyll serve
Running Jekyll XXX but site expected 3.0.3 see http://jekyllrb.com/docs/upgrading/2-to-3/ for potential changes
```

Running the correct version would print nothing

```bash
$ jekyll _3.0.3_ serve
...
```

This matches a feature on CloudCannon at http://docs.cloudcannon.com/building/jekyll/.

**Comment by envygeeks** [comment]
Why are we trying to take the place of a Gemfile.lock and Gemfile?

**Comment by parkr** [comment]
@envygeeks This reminds me of the conversation that was had about the `Gemfile.lock` `BUNDLED WITH` statement. It's all about spotting problems because the activated version is different than the one you have used previously or want to use.

Pinning your site to a Jekyll version not descriptively (Gemfile), but prescriptively is a great idea. This is telling your site "let me know if you are expecting a different Jekyll version", whereas Gemfile is describing the Jekyll version that will be used. If these two things are used together, it will surface for folks the case where they have upgraded their Jekyll version but haven't said their site is compatible with it.

**Comment by benbalter** [comment]
That's a cool idea. What do you think about supporting pessimistic operators? E.g., I don't care if I'm running a different patch release, and wouldn't want to have to change a version controlled file every time I run bundle update.

**Comment by envygeeks** [comment]
And what @benbalter just said is why I made my comment.

**Comment by envygeeks** [comment]
That said, the Gemfile exists for a reason, the user can do what @benbalter just suggested and that's exactly what it was designed to do.  The user can loosen or tighten their dependencies as much as they wish.  What benefit would this provide over that other than adding extra strain onto the management of a site... that is, unless you are start tacking on more features like `jekyll update version` (or any derivative) which again why are we taking over bundlers job?

People might disagree with me, but I don't see how this solves the problem of educating users.

**Comment by mattr-** [comment]
I'm not in favor of this idea either. I'd rather see users make use of pessimistic versioning in the Gemfile they they're using to pull Jekyll. IMHO, that's where this functionality belongs. 

IMHO, if the `github-pages` gem wants to implement this sort of check, similar to how cloud cannon apparently does it, then i'm good with that. I don't think this belongs in Jekyll proper. 

My :two: :moneybag: worth.

**Comment by parkr** [comment]
This comes from a place of disliking getting the "you're using a different version than you started with and you didn't know that" issues. I think we change things enough to not allow pessimistic operators in this check. If it says "3.1.1" and you're using 3.1.2, then notify them. We get a lot of Ruby newbies and this might go a long way to surfacing issues.

> IMHO, if the github-pages gem wants to implement this sort of check, similar to how cloud cannon apparently does it, then i'm good with that. I don't think this belongs in Jekyll proper.

/cc https://github.com/github/pages-gem/pull/209

**Comment by GeorgePhillips** [comment]
I think this would be useful as many users install Jekyll as a binary without a Gemfile. This is the method suggested by the front page of http://jekyllrb.com/. Imagine this onboarding flow for a new user:

- Installs Jekyll with `gem install jekyll`
- Runs `jekyll new` for their first site, it works perfectly
- Downloads a template from somewhere built for 2.4.0
- Runs jekyll serve on the new template and half the page is missing due to the change in collections

The user never got introduced to Gemfiles and is unaware of difference between the two sites. Jekyll or the template simply appear broken.

**Comment by parkr** [comment]
@GeorgePhillips I agree; I think we need to rethink that and propose the Gemfile + Bundler route upfront. Bundler makes working in Ruby so much easier.

**Comment by envygeeks** [comment]
Where are these users getting these binaries from? As far as the bundler argument, yeah we are pushing the documentation t owards that direction already by shifting things from `gem` this to `bundle that`

**Comment by GeorgePhillips** [comment]
most new users install with `gem install jekyll` which installs it to the PATH for use with `jekyll serve`. I consider that as a binary.

**Comment by envygeeks** [comment]
That's not a binary though, that's the same source you get when you bundle install.  I just need to get that out of the way before users get confused into thinking that we actually offer binaries of which we don't.

**Comment by GeorgePhillips** [comment]
Apologies for the terminology mix up

**Comment by GeorgePhillips** [comment]
This could be achieved with the regex `/^(\d+|x)\.(\d+\.(\d+|x)|x)$/` which would match the following:

```
3.x
2.4.0
2.4.x
2.5.3
3.0.x
3.1.1
```

**Comment by envygeeks** [comment]
"3.x" is the antithesis of the Ruby way and will confuse our users even more.

**Comment by GeorgePhillips** [comment]
`/^(\d+)\.(\d+)\.(\d+|x)$/`

**Comment by envygeeks** [comment]
Let me rephrase that, any version that involves ".x" is the antithesis of Ruby.  In Ruby it's "~> 3.0", ">=", "<=", "=" or just "3.0" if your goal is to educate users (and I still don't support this feature at all) then don't encourage the opposite of the Ruby way.

**Comment by GeorgePhillips** [comment]
We might have gone off topic. I was simply trying to provide possible solutions to the problem. Sadly my solution cannot match a Gemfile as Gemfiles are not YAML. Base Jekyll requires very little Ruby knowledge, in fact building Jekyll sites is mostly YAML, HTML, Markdown and Liquid. Should we not try to match that rather than the language that Jekyll is built in?

**Comment by GeorgePhillips** [comment]
This follows the same lines of reasoning as @benbalter with https://github.com/jekyll/jekyll/issues/4509.

> The idea being, if you're not from the Ruby world, "gems" isn't a very intuitive term (even though that's what they are). Further, users shouldn't have to care about (or see) the behind-the-scenes plumbing, in terms of how things are distributed.

**Comment by mattr-** [comment]
> I think we change things enough to not allow pessimistic operators in this check. If it says "3.1.1" and you're using 3.1.2, then notify them.

This idea that we change enough things that we have to worry about patch level version numbers scares me. It seems wrong to me – based solely on how i like to interpret SemVer – that something should break so severely with a patch or a minor release that we'd need to warn users about it. I know that SemVer was written with library APIs in mind. In the long run though, I think we'll be better off following those same principles when it comes to our CLI and our generation process such that we only introduce major breaks with a major version number bump. Doing so would remove the need for a version restriction in the config, IMHO.

**Comment by envygeeks** [comment]
> This idea that we change enough things that we have to worry about patch level version numbers scares me. It seems wrong to me – based solely on how i like to interpret SemVer – that something should break so severely with a patch or a minor release that we'd need to warn users about it. I know that SemVer was written with library APIs in mind. In the long run though, I think we'll be better off following those same principles when it comes to our CLI and our generation process such that we only introduce major breaks with a major version number bump. Doing so would remove the need for a version restriction in the config, IMHO.

Agreed, in whole.

**Comment by GeorgePhillips** [comment]
@mattr- I agree about the need to include patch level version numbers. If semantic versioning is used correctly only major/minor versions could mean different output from a build. Do you not agree that immediate warning of mismatched major/minor versions could save debugging time to many users?

**Comment by mattr-** [comment]
@GeorgePhillips No, I do not agree. A minor version bump should not affect my build in any meaningful way in terms of the output generated. IOW, moving from Jekyll 3.1 to 3.2 should not break my site and there should be no reason for me to even worry about it breaking my site.

**Comment by chrisfinazzo** [comment]
@GeorgePhillips Note that #4358, #4368, and #4391 all deal with moving to Bundler + a Gemfile as one of the preferred installation methods. As you can probably tell, I'm a big fan of this approach 😀.

On this specific issue, I side with what @mattr- and @envygeeks have said - unless there is a major version change, this just creates an unnecessary hurdle to jump in order to get things working. The [upgrade][] [guides][] exist in the form that they do because these are the most common transitions where people are likely to run into problems.

As far as whether this approach is something that would work in a GitHub Pages environment, I think it has potential, but I would defer to @parkr regarding the details of how that should be handled.

[upgrade]: http://jekyllrb.com/docs/upgrading/0-to-2/

[guides]: http://jekyllrb.com/docs/upgrading/2-to-3/

**Comment by parkr** [comment]
> @GeorgePhillips: If semantic versioning is used correctly only major/minor versions could mean different output from a build
> ...
> @mattr-: A minor version bump should not affect my build in any meaningful way in terms of the output generated.

FWIW, this doesn't always happen. Bug fixes often introduce changes that intentionally modify the output of a site. Bug fixes are often issued because the site isn't being generated properly and therefore _must_ be changed. If you started your site with a broken version of Jekyll, then you'd necessarily see changes in the next patch version.

We have not been good about not changing the output of a site with new versions. This is largely because we're still working to root out inconsistencies in various features, but it's the reality.

---

Thanks for the great discussion everyone, though it got unnecessarily negative. I'd like us to be welcoming differing opinions rather than shutting them down. On the whole, we should be asking "why" more, instead of jumping to conclusions. I struggle with this every day. Sadly, it's arguments like these that can send people walking away from our community rather than running back with more suggestions. 

A message like the proposed might aid someone in their workflow quite a bit ("hm, why is my site broken today but was working yesterday?") but might just be annoying to others. A classic differing of opinion in which an "opt-in" strategy might be of most use. That opt-in strategy could be by turning off the feature unless the user adds the configuration option (like we do most things), or through the use of a plugin.

**Comment by envygeeks** [comment]
> Thanks for the great discussion everyone, though it got unnecessarily negative. I'd like us to be welcoming differing opinions rather than shutting them down.

This statement is ironic and contradicts itself on the face of what it's trying to do.  Just sayin' :wink:

**Comment by TWiStErRob** [comment]
(I didn't read past @mattr-'s first comment) I think this would be extremely useful for users of pages-gem, as they don't follow semver (-> hard to reference "only minor changes" in Gemfile). Suppose my Gemfile only contains: `gem 'github-pages', '>= 40'` where it's unclear which Jekyll is pulled in, and I don't really care about it as long as it builds my site (I guess 2.x.x). The problem comes when v44 was released and it silently just breaks my build, because of a major change.

**Comment by TWiStErRob** [comment]
I don't that PR fixes this issue.
@benbalter @parkr so how can #4542 prevent users from breaking their site? How can we use this prevent github-pages from updating a major version of jekyll? I have an existing site, which already has a Gemfile with a line `gem github-pages`.

**Comment by benbalter** [comment]
@TWiStErRob the GitHub Pages Gem is purposely not semantically versioned. It's intended to help users replicate the GitHub Pages production environment as closely as possible on their local machine. As such, if GitHub Pages updates to version 2, and you're locked on version 1, it would defeat the purpose of the Gem (and you should bundle Jekyll and desired dependencies locally). Since https://github.com/github/pages-gem/pull/209 has landed, we've been experimenting with a plugin that would warn you if your version of the GitHub Page Gem mismatches with what's being used in production (thus suggesting you need to run `bundle update`), which might solve the problem you're describing.
